string_match_without_scan_or_regex.rb

```diff
@@ -1,36 +1,63 @@
-def match?(string, substring)
-  string_array = string.split("")
-  substring_array = substring.split("")
-  character_matches = []
-  substring_array.each_with_index do |substring_character, substring_index|
-    find_character_match(string_array, substring_array, substring_character, substring_index, character_matches)
+class String
+
+  def match?(substring)
+    if substring.class == String
+      #passing hash to collect matches, as well as the starting
+      #index 0 for the recursive call
+      matches = generate_matches(self, substring, {}, 0)
+      matches.has_value?(substring)
+    else
+      wrong_argument_type
+    end
   end
-  character_matches.join("") == substring ? true : false
-end
 
-def find_character_match(string_array, substring_array, substring_character, substring_index, character_matches)
-  if string_array.include?(substring_character)
-    if substring_index == 0
-      character_matches.push(substring_character)
-    elsif substring_index > 0
-      string_array.each_with_index do |string_character, string_index|
-
-        characters_do_match = string_character == substring_character
-        previous_string_index_not_negative = string_index - 1 > -1
-        previous_characters_do_match = string_array[string_index - 1] == substring_array[substring_index - 1]
-
-        if characters_do_match && previous_string_index_not_negative && previous_characters_do_match
-          character_matches.push(substring_character)
-          break
-        end
-      end
+  def match_count(substring)
+    if substring.class == String
+      #passing hash to collect matches, as well as the starting
+      #index 0 for the recursive call
+      matches = generate_matches(self, substring, {}, 0)
+      matches.length
+    else
+      wrong_argument_type
     end
   end
-end
 
-p match?("likes", "like") == true
-p match?("muchlikes", "like") == true
-p match?("ikelmuchlikes", "like") == true
-p match?("bass1lurv", " 1lurv") == false
-p match?("1basslurv", "1lurv") == false
-p match?("ikemarl", "like") == false
\ No newline at end of file
+  def match_indices(substring)
+    if substring.class == String
+      #passing hash to collect matches, as well as the starting
+      #index 0 for the recursive call
+      matches = generate_matches(self, substring, {}, 0)
+      matches.keys
+    else
+      wrong_argument_type
+    end
+  end
+
+  private
+
+  #used as a constant for wrong argument type
+  def wrong_argument_type
+    raise 'wrong argument type (expected String)'
+  end
+
+
+  #for first pass, collection_of_matches should be {} and match_index should be 0
+  def generate_matches(string, substring, collection_of_matches, match_index)
+
+    string_slice = string.slice(0, substring.length)
+
+    collection_of_matches[match_index] = string_slice if string_slice == substring
```

Comment:
Also, is there a data structure you can use which doesn't require indices?  This would allow you to get rid of `match_index`, which is really redundant with your collection count.